WordPress/src/main/java/org/wordpress/android/viewmodel/giphy/provider/TenorProvider.kt

```diff
@@ -50,45 +57,78 @@ internal class TenorProvider @JvmOverloads constructor(
      * or that a direct failure was found by Tenor API
      *
      * Method is inlined for better high-order functions performance
      */
-    private inline fun IApiClient.simpleSearch(
+    private inline fun IApiClient.enqueueSearchRequest(
         query: String,
         position: String,
         loadSize: Int?,
         crossinline onSuccess: (GifsResponse) -> Unit,
         crossinline onFailure: (Throwable?) -> Unit
-    ) {
-        search(
-                ApiClient.getServiceIds(context),
-                query,
-                loadSize.fittedToMaximumAllowed,
-                position,
-                MediaFilter.BASIC,
-                AspectRatioRange.ALL
-
-        ).enqueue(object : WeakRefCallback<Context, GifsResponse>(context) {
+    ) = buildSearchCall(query, loadSize, position).apply {
+        enqueue(object : WeakRefCallback<Context, GifsResponse>(context) {
             override fun success(context: Context, response: GifsResponse?) {
+                this@apply.cancel()
                 response?.let(onSuccess)
-                        ?: onFailure(
-                                GifRequestFailedException(
-                                        context.getString(string.giphy_picker_empty_search_list)
-                                )
-                        )
+                        ?: onFailure(buildGifRequestFailure(string.giphy_picker_empty_search_list))
             }
 
             override fun failure(context: Context, throwable: Throwable?) {
+                this@apply.cancel()
                 onFailure(throwable)
             }
         })
+
+        dispatchTimeoutClock(onFailure)
+    }
+
+    /**
+     * This method act as a simplification to call a search within the Tenor API and the callback
+     * creation
+     *
+     * [MediaFilter] must be BASIC or the returned Media will not have a displayable thumbnail
+     * All other provided parameters are set following the Tenor API Documentation
+     */
+    private fun IApiClient.buildSearchCall(
+        query: String,
+        loadSize: Int?,
+        position: String
+    ) = search(
+            ApiClient.getServiceIds(context),
+            query,
+            loadSize.fittedToMaximumAllowed,
+            position,
+            MediaFilter.BASIC,
+            AspectRatioRange.ALL
+    )
+
+    /**
+     * This method counts from 0 to [DEFAULT_SECONDS_TO_TIMEOUT] to make sure that
+     * every request responds within the expected amount of time.
+     *
+     * If the call isn't canceled until the timer reaches the limit,
+     * the [onFailure] will be called passing an [GifRequestTimeoutException]
+     *
+     * But if the call is canceled before the timer reaches the limit,
+     * the coroutine job will right away be canceled and the [onFailure] call
+     * won't be reached
+     *
+     * Method is inlined for better high-order functions performance
+     */
+    private inline fun Call<GifsResponse>.dispatchTimeoutClock(
```

Comment:
This is also way too complicated tbh. If we want to implement a timeout, we can do something like this:
```
        val timeoutJob = scope.launch {
            delay(DEFAULT_SECONDS_TO_TIMEOUT * ONE_SECOND_IN_MILLIS)
             buildSearchCall.cancel()
        }
```
and do something like this on the onSuccess/onFailure callbacks:
```

                if (timeoutJob.isActive) {
                    timeoutJob.cancel()
                }
```
I've also looked into the ApiService builder and there is actually timeout built in: 
`com.tenor.android.core.network.ApiService.IBuilder#timeout`. I think we should use that instead.

Reply:
I decided to investigate further the implementation options and how the Tenor SDK was making use of the retrofit calls, and the solution I've found made the `dispatchTimeoutClock` unnecessary, so I completely removed it. The fix can be found at f6123f6993b77a68deeec3db38576f2024c4b000. I've also added at the bottom a final comment given further context about my solution and how I decided to proceed.